File: hackerrank/typescript/src/utopian-tree.ts
/**
 * @param n - Number of growth cycles
 * @returns Height after growth
 *
 * Challenge:
 * https://www.hackerrank.com/challenges/utopian-tree/problem
 */
export const utopianTree = (n: number): number => {
	let h = 1
	for (let index = 1; index <= n; index++) {
		if (index % 2 == 0) {
			h++
		} else {
			h *= 2
		}
	}
	return h
}

if (import.meta.vitest) {
	const { it, expect } = import.meta.vitest

	it('Utopian Tree', () => {
		expect(utopianTree(5)).toBe(14)
	})
}

File: hackerrank/rust/src/utopian_tree.rs
#[allow(dead_code)]
/// Calculate the height of tree after `n` growth cycles
///
/// * `n`: number of growth cycles
///
/// Challenge:
/// https://www.hackerrank.com/challenges/utopian-tree/problem

fn utopian_tree(n: i32) -> i32 {
    let mut h = 1;
    for i in 1..=n {
        if i % 2 != 0 {
            h *= 2;
        } else {
            h += 1;
        }
    }
    h
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_one() {
        assert_eq!(utopian_tree(5), 14);
    }
}
